hw/opentitan/ot_otp_dj.c

```diff
@@ -1522,100 +1437,117 @@ static uint64_t ot_otp_dj_get_buffered_part_digest(OtOTPDjState *s, int part)
 
 static bool ot_otp_dj_is_part_digest_offset(int part, hwaddr addr)
 {
-    uint16_t offset = ot_otp_dj_get_part_digest_offset(part);
+    uint16_t offset = OtOTPPartDescs[part].digest_offset;
 
     return (offset != UINT16_MAX) && ((addr & ~OTP_DIGEST_ADDR_MASK) == offset);
 }
 
+static bool ot_otp_dj_is_part_zer_offset(int part, hwaddr addr)
+{
+    uint16_t offset = OtOTPPartDescs[part].zer_offset;
+
+    return (offset != UINT16_MAX) && ((addr & ~OTP_ZER_ADDR_MASK) == offset);
+}
+
+static uint32_t ot_otp_dj_get_part_digest_reg(OtOTPDjState *s, uint32_t offset)
+{
+    /*
+     * Offset it the register offset from the first read digest 32-bit register.
+     * all digest are 64-bits, which means each partitition have two 32-bit
+     * registers to expose their digest.
+     *
+     * Not all partitions have digests, which means the offset argument is the
+     * relative partition offset for those partitions that features a digest, as
+     * there is no defined digest access registers defined for partitions that
+     * do not have a digest.
+     *
+     * Need to traverse the partition table to only account for those partitions
+     * with a digest to match the proper offset.
+     */
+    unsigned part_look_ix = (int)offset >> 1u;
+    bool hi = (bool)(offset & 0x1);
+
+    unsigned part_ix = 0;
+    for (unsigned part_dig_ix = 0; part_ix < OTP_PART_COUNT; part_ix++) {
+        if (ot_otp_dj_has_digest(part_ix)) {
+            if (part_dig_ix == part_look_ix) {
+                break;
+            }
+            part_dig_ix++;
+        }
+    }
```

Comment:
I think the shifting for the double registers makes it harder to follow, IMO I think it would be easier to understand (and more general) if we do:

```c
#define OTP_DIGEST_REG_COUNT (sizeof(uint64_t) / sizeof(uint32_t))
```

and then:

```suggestion
    for (unsigned part_ix = 0; part_ix < OTP_PART_COUNT; part_ix++) {
        if (ot_otp_dj_has_digest(part_ix)) {
            if (part_ix <= offset && offset < (part_ix + OTP_DIGEST_REG_COUNT)) {
                break;
            }
            part_dig_ix += OTP_DIGEST_REG_COUNT;
        }
    }
```

or similar. But feel free to keep doing it this way if you prefer.

(aside: thanks for the doc comment, it was a big help to understand this part)

Reply:
It seems harder for me to read. We would perform an addition of a partition index with a register count ... I'm a bit lost here.

Reply:
Sorry, there was a problem in my original suggestion, I was evidently too tired when I wrote it. I meant to check `part_dig_ix` in the conditional i.e.:
```c
if (part_dig_ix <= offset && offset < (part_dig_ix + OTP_DIGEST_REG_COUNT)) {
```
The partition index should always remain correct, but `part_dig_ix` (would then maybe be renamed to something like `part_dig_reg_offset` or something) would then track the offset explicitly by summing the register counts of digest partitions encountered so far.

If you find it less readable then definitely just leave it as is, however the main part of the original logic I struggled to follow was that
```c
unsigned part_look_ix = (int)offset >> 1u;
```
was responsible for considering the "2 32-bit registers per digest" part, this would be much more transparent to me (and generalized) with something like:
```
unsigned part_look_ix = offset / OTP_DIGEST_REG_COUNT;
```
which I imagine would probably compile down the same anyway?

Reply:
Oh ok, I'll change that. Definitely more readable.

Reply:
Actually, there's already a `NUM_DIGEST_WORDS` generated by OT tooling.

Reply:
Would it be easier to understand with:
````diff
diff --git a/hw/opentitan/ot_otp_dj.c b/hw/opentitan/ot_otp_dj.c
index 8a51f047f1..7ec0905a75 100644
--- a/hw/opentitan/ot_otp_dj.c
+++ b/hw/opentitan/ot_otp_dj.c
@@ -1464,15 +1464,33 @@ static uint32_t ot_otp_dj_get_part_digest_reg(OtOTPDjState *s, uint32_t offset)
      * Need to traverse the partition table to only account for those partitions
      * with a digest to match the proper offset.
      */
-    unsigned part_look_ix = (int)offset >> 1u;
-    bool hi = (bool)(offset & 0x1);
 
+    /*
+     * part_look_ix is the index of the partition in the contiguous array of
+     * digest registers, which would be equivalent as the index of the partition
+     * that would exist in a virtual partition-with-digest array.
+     */
+    unsigned part_look_ix = (int)(offset / NUM_DIGEST_WORDS);
+    /* whether to retrieve the top most 32-bit of the digest or not */
+    bool hi = (bool)(offset & 0x1u);
+
+    /* part_ix: the partition number in the global partition array */
     unsigned part_ix = 0;
+    /* traverse the partition array and count each partition with a digest */
     for (unsigned part_dig_ix = 0; part_ix < OTP_PART_COUNT; part_ix++) {
         if (ot_otp_dj_has_digest(part_ix)) {
             if (part_dig_ix == part_look_ix) {
+                /*
+                 * break as soon as the partition-with-digest index matches
+                 * the one defined from the offset argument; part_ix contains
+                 * the actual partition index we are looking for.
+                 */
                 break;
             }
+            /*
+             * increment the partition-with-digest index only if the current
+             * partition feature a digest
+             */
             part_dig_ix++;
         }
     }
@@ -1480,8 +1498,11 @@ static uint32_t ot_otp_dj_get_part_digest_reg(OtOTPDjState *s, uint32_t offset)
     /*
      * Know for sure the latest partition (LC) does not have a digest, so if
      * the part_ix as reached the latest partition, there is something wrong
-     * with the partition table or the register definitions
+     * with the partition table or the register definitions.
+     * It is assumed that LifeCycle partition is the last partition.
      */
+    static_assert(OTP_PART_LIFE_CYCLE == NUM_PART - 1,
+                  "LC is expected to be the last partition");
     g_assert(part_ix < OTP_PART_COUNT);
 
     const OtOTPPartDesc *part = &OtOTPPartDescs[part_ix];
````

Reply:
Yes that is very clear. The changes & comments (especially on `part_look_ix` and `part_ix`) are very helpful. I think maybe you could replace the two comments in the `ot_otp_dj_has_digest(part_ix)` branch with just one shorter comment in the branch like:
```
/*
 * stop searching if we've found the part-with-digest defined from the offset
 * argument. Otherwise, increment the part-with-digest index and continue.
/* 
```

Reply:
Done.